File: docs/reference/v4_simulator/reeds-shepp.js
/* Reeds–Shepp analytic families adapted from PythonRobotics, MIT.
 * Atsushi Sakai and Videh Patel. See THIRD_PARTY_NOTICES.txt.
 * Twelve base families × time reversal × reflection. Every accepted endpoint
 * is independently checked by bicycle integration before collision validation.
 */
(function(G){'use strict';const C=G.CarbotCore,pi=Math.PI,h=pi/2,m=C.wrap,polar=(x,y)=>[Math.hypot(x,y),Math.atan2(y,x)],sin=Math.sin,cos=Math.cos,sqrt=Math.sqrt,acos=x=>Math.acos(C.clamp(x,-1,1)),asin=x=>Math.asin(C.clamp(x,-1,1));
const F=[
 (x,y,p)=>{let[u,t]=polar(x-sin(p),y-1+cos(p)),v=m(p-t);if(t>=0&&t<=pi&&v>=0&&v<=pi)return[[t,u,v],'LSL']},
 (x,y,p)=>{let[r,b]=polar(x+sin(p),y-1-cos(p));if(r*r>=4){let u=sqrt(r*r-4),t=m(b+Math.atan2(2,u)),v=m(t-p);if(t>=0&&v>=0)return[[t,u,v],'LSR']}},
 (x,y,p)=>{let[r,b]=polar(x-sin(p),y-1+cos(p));if(r<=4){let a=acos(r/4),t=m(a+b+h),u=m(pi-2*a),v=m(p-t-u);return[[t,-u,v],'LRL']}},
 (x,y,p)=>{let[r,b]=polar(x-sin(p),y-1+cos(p));if(r<=4){let a=acos(r/4),t=m(a+b+h),u=m(pi-2*a),v=m(-p+t+u);return[[t,-u,-v],'LRL']}},
 (x,y,p)=>{let[r,b]=polar(x-sin(p),y-1+cos(p));if(r>1e-8&&r<=4){let u=acos(1-r*r/8),a=asin(2*sin(u)/r),t=m(-a+b+h),v=m(t-u-p);return[[t,u,-v],'LRL']}},
 (x,y,p)=>{let[r,b]=polar(x+sin(p),y-1-cos(p));if(r<=2){let a=acos((r+2)/4),t=m(b+a+h),u=m(a),v=m(p-t+2*u);if(t>=0&&u>=0&&v>=0)return[[t,u,-u,-v],'LRLR']}},
 (x,y,p)=>{let[r,b]=polar(x+sin(p),y-1-cos(p)),u2=(20-r*r)/16;if(u2>=0&&u2<=1&&r>1e-8){let u=acos(u2),a=asin(2*sin(u)/r),t=m(b+a+h),v=m(t-p);if(t>=0&&v>=0)return[[t,-u,-u,v],'LRLR']}},
 (x,y,p)=>{let[r,b]=polar(x-sin(p),y-1+cos(p));if(r>=2){let q=sqrt(r*r-4),u=q-2,a=Math.atan2(2,q),t=m(b+a+h),v=m(t-p+h);if(t>=0&&v>=0)return[[t,-h,-u,-v],'LRSL']}},
 (x,y,p)=>{let[r,b]=polar(x+sin(p),y-1-cos(p));if(r>=2){let t=m(b+h),u=r-2,v=m(p-t-h);if(t>=0&&v>=0)return[[t,-h,-u,-v],'LRSR']}},
 (x,y,p)=>{let[r,b]=polar(x-sin(p),y-1+cos(p));if(r>=2){let q=sqrt(r*r-4),u=q-2,a=Math.atan2(q,2),t=m(b-a+h),v=m(t-p-h);if(t>=0&&v>=0)return[[t,u,h,-v],'LSRL']}},
 (x,y,p)=>{let[r,b]=polar(x+sin(p),y-1-cos(p));if(r>=2){let t=m(b),u=r-2,v=m(p-t-h);if(t>=0&&v>=0)return[[t,u,h,-v],'LSLR']}},
 (x,y,p)=>{let[r,b]=polar(x+sin(p),y-1-cos(p));if(r>=4){let q=sqrt(r*r-4),u=q-4,a=Math.atan2(2,q),t=m(b+a+h),v=m(t-p);if(t>=0&&v>=0)return[[t,-h,-u,-h,v],'LRSLR']}}
];
function candidates(start,goal,r,step=.006){let local=C.toLocal(start,goal),x=local.x/r,y=local.y/r,phi=C.wrap(goal.a-start.a),out=[];for(let f of F)for(let variant=0;variant<4;variant++){let flip=variant===1||variant===3,reflect=variant>=2,q=f(flip?-x:x,reflect?-y:y,(flip!==reflect)?-phi:phi);if(!q)continue;let lengths=q[0].map(v=>v*r*(flip?-1:1)),types=q[1].split('').map(v=>reflect?(v==='L'?'R':v==='R'?'L':'S'):v),p={...start},path=[],cost=0;for(let j=0;j<lengths.length;j++){let d=lengths[j];if(Math.abs(d)<1e-8)continue;let dir=Math.sign(d),k=types[j]==='L'?1/r:types[j]==='R'?-1/r:0,n=Math.ceil(Math.abs(d)/step);if(!path.length)path.push({...p,dir,k});else path.push({...p,dir,k});let origin={...p};for(let i=1;i<=n;i++){p=C.bicycle(origin,d*i/n,k);path.push({...p,dir,k})}cost+=Math.abs(d)}if(!path.length||Math.hypot(p.x-goal.x,p.y-goal.y)>1e-5||Math.abs(C.wrap(p.a-goal.a))>1e-5)continue;out.push({path,cost,types:types.join(''),lengths,variant})}out.sort((a,b)=>a.cost-b.cost);return out}
function plan(start,goal,course,c){let all=candidates(start,goal,C.geometry(c).r);for(const q of all){q.valid=q.path.every(p=>course.bodyClear(p,c,.003));q.reject=q.valid?'':'Footprint leaves drivable area'}let winner=all.find(q=>q.valid);return winner?{...winner,expanded:48,reason:'Live Reeds–Shepp connection',candidates:all.length,evaluated:all}:{path:[],expanded:48,candidates:all.length,evaluated:all,reason:'No collision-free analytic connection'}}
G.CarbotRS={candidates,plan};
})(globalThis);

File: src/carbot_planning/carbot_planning/reeds_shepp.py
"""Reeds-Shepp connections (no ROS): faithful port of V4 reeds-shepp.js.

V4: "Reeds-Shepp analytic families adapted from PythonRobotics, MIT. Atsushi
Sakai and Videh Patel. Twelve base families x time reversal x reflection.
Every accepted endpoint is independently checked by bicycle integration
before collision validation."

Used by block 11 (parking) and block 12 (recovery). Verified against the JS in
Node (test_reeds_shepp.py, tools/v4_harness/rs_v4.js).

Path points: N x 5 array [x, y, yaw, dir, k] (dir +1 forward / -1 reverse,
k = curvature of the segment the point belongs to). Every segment starts with
a copy of the previous segment's last pose carrying the new dir / k, exactly
as V4 pushes it, so a cusp shows up as two identical poses with different dir.
"""
import math
from dataclasses import dataclass, field
from typing import Callable, List, Optional, Tuple

import numpy as np

PI = math.pi
H = PI / 2


def wrap(a: float) -> float:
    return math.atan2(math.sin(a), math.cos(a))


def _clamp(x: float, a: float, b: float) -> float:
    return max(a, min(b, x))


def _acos(x: float) -> float:
    return math.acos(_clamp(x, -1.0, 1.0))


def _asin(x: float) -> float:
    return math.asin(_clamp(x, -1.0, 1.0))


def _polar(x: float, y: float) -> Tuple[float, float]:
    return math.hypot(x, y), math.atan2(y, x)


# --------------------------------------------------------------------------- the 12 families (V4 order)
def _f1(x, y, p):
    u, t = _polar(x - math.sin(p), y - 1 + math.cos(p))
    v = wrap(p - t)
    if 0 <= t <= PI and 0 <= v <= PI:
        return [t, u, v], 'LSL'


def _f2(x, y, p):
    r, b = _polar(x + math.sin(p), y - 1 - math.cos(p))
    if r * r >= 4:
        u = math.sqrt(r * r - 4)
        t = wrap(b + math.atan2(2, u))
        v = wrap(t - p)
        if t >= 0 and v >= 0:
            return [t, u, v], 'LSR'


def _f3(x, y, p):
    r, b = _polar(x - math.sin(p), y - 1 + math.cos(p))
    if r <= 4:
        a = _acos(r / 4)
        t = wrap(a + b + H)
        u = wrap(PI - 2 * a)
        v = wrap(p - t - u)
        return [t, -u, v], 'LRL'


def _f4(x, y, p):
    r, b = _polar(x - math.sin(p), y - 1 + math.cos(p))
    if r <= 4:
        a = _acos(r / 4)
        t = wrap(a + b + H)
        u = wrap(PI - 2 * a)
        v = wrap(-p + t + u)
        return [t, -u, -v], 'LRL'


def _f5(x, y, p):
    r, b = _polar(x - math.sin(p), y - 1 + math.cos(p))
    if 1e-8 < r <= 4:
        u = _acos(1 - r * r / 8)
        a = _asin(2 * math.sin(u) / r)
        t = wrap(-a + b + H)
        v = wrap(t - u - p)
        return [t, u, -v], 'LRL'


def _f6(x, y, p):
    r, b = _polar(x + math.sin(p), y - 1 - math.cos(p))
    if r <= 2:
        a = _acos((r + 2) / 4)
        t = wrap(b + a + H)
        u = wrap(a)
        v = wrap(p - t + 2 * u)
        if t >= 0 and u >= 0 and v >= 0:
            return [t, u, -u, -v], 'LRLR'


def _f7(x, y, p):
    r, b = _polar(x + math.sin(p), y - 1 - math.cos(p))
    u2 = (20 - r * r) / 16
    if 0 <= u2 <= 1 and r > 1e-8:
        u = _acos(u2)
        a = _asin(2 * math.sin(u) / r)
        t = wrap(b + a + H)
        v = wrap(t - p)
        if t >= 0 and v >= 0:
            return [t, -u, -u, v], 'LRLR'


def _f8(x, y, p):
    r, b = _polar(x - math.sin(p), y - 1 + math.cos(p))
    if r >= 2:
        q = math.sqrt(r * r - 4)
        u = q - 2
        a = math.atan2(2, q)
        t = wrap(b + a + H)
        v = wrap(t - p + H)
        if t >= 0 and v >= 0:
            return [t, -H, -u, -v], 'LRSL'


def _f9(x, y, p):
    r, b = _polar(x + math.sin(p), y - 1 - math.cos(p))
    if r >= 2:
        t = wrap(b + H)
        u = r - 2
        v = wrap(p - t - H)
        if t >= 0 and v >= 0:
            return [t, -H, -u, -v], 'LRSR'


def _f10(x, y, p):
    r, b = _polar(x - math.sin(p), y - 1 + math.cos(p))
    if r >= 2:
        q = math.sqrt(r * r - 4)
        u = q - 2
        a = math.atan2(q, 2)
        t = wrap(b - a + H)
        v = wrap(t - p - H)
        if t >= 0 and v >= 0:
            return [t, u, H, -v], 'LSRL'


def _f11(x, y, p):
    r, b = _polar(x + math.sin(p), y - 1 - math.cos(p))
    if r >= 2:
        t = wrap(b)
        u = r - 2
        v = wrap(p - t - H)
        if t >= 0 and v >= 0:
            return [t, u, H, -v], 'LSLR'


def _f12(x, y, p):
    r, b = _polar(x + math.sin(p), y - 1 - math.cos(p))
    if r >= 4:
        q = math.sqrt(r * r - 4)
        u = q - 4
        a = math.atan2(2, q)
        t = wrap(b + a + H)
        v = wrap(t - p)
        if t >= 0 and v >= 0:
            return [t, -H, -u, -H, v], 'LRSLR'


FAMILIES: List[Callable] = [_f1, _f2, _f3, _f4, _f5, _f6, _f7, _f8, _f9, _f10, _f11, _f12]


# --------------------------------------------------------------------------- sampling
def _segment(x: float, y: float, a: float, d: float, k: float, n: int) -> np.ndarray:
    """core.js bicycle(origin, d*i/n, k) for i = 1..n (vectorised, same float ops)."""
    s = d * np.arange(1, n + 1, dtype=float) / n
    if abs(k) < 1e-9:
        return np.column_stack([x + s * math.cos(a), y + s * math.sin(a), np.full(n, a)])
    a2 = a + s * k
    return np.column_stack([x + (np.sin(a2) - math.sin(a)) / k, y + (-np.cos(a2) + math.cos(a)) / k,
                            np.arctan2(np.sin(a2), np.cos(a2))])


@dataclass
class RSPath:
    points: np.ndarray          # N x 5 [x, y, a, dir, k]
    cost: float                 # total |length|
    types: str                  # e.g. 'LSR'
    lengths: List[float]        # signed segment lengths (m)
    variant: int
    valid: bool = False
    reject: str = ''
    extra: dict = field(default_factory=dict)

    @property
    def reverse_length(self) -> float:
        return float(sum(-d for d in self.lengths if d < 0))


def candidates(start, goal, r: float, step: float = 0.006) -> List[RSPath]:
    """V4 CarbotRS.candidates: every family x {plain, flip, reflect, both}, sampled at
    `step`, endpoint re-checked (1e-5 m / rad), sorted by length (stable, as V8)."""
    sx, sy, sa = start
    gx, gy, ga = goal
    c, s = math.cos(sa), math.sin(sa)
    lx, ly = (gx - sx) * c + (gy - sy) * s, -(gx - sx) * s + (gy - sy) * c
    x, y, phi = lx / r, ly / r, wrap(ga - sa)
    out: List[RSPath] = []
    for f in FAMILIES:
        for variant in range(4):
            flip = variant in (1, 3)
            reflect = variant >= 2
            q = f(-x if flip else x, -y if reflect else y, -phi if flip != reflect else phi)
            if not q:
                continue
            lengths = [v * r * (-1 if flip else 1) for v in q[0]]
            types = [('R' if t == 'L' else 'L' if t == 'R' else 'S') if reflect else t for t in q[1]]
            px, py, pa = sx, sy, sa
            parts = []
            cost = 0.0
            for d, t in zip(lengths, types):
                if abs(d) < 1e-8:
                    continue
                dr = 1.0 if d > 0 else -1.0
                k = 1 / r if t == 'L' else -1 / r if t == 'R' else 0.0
                n = int(math.ceil(abs(d) / step))
                seg = _segment(px, py, pa, d, k, n)
                parts.append(np.array([[px, py, pa, dr, k]]))
                parts.append(np.column_stack([seg, np.full(n, dr), np.full(n, k)]))
                px, py, pa = seg[-1]
                cost += abs(d)
            if not parts or math.hypot(px - gx, py - gy) > 1e-5 or abs(wrap(pa - ga)) > 1e-5:
                continue
            out.append(RSPath(np.vstack(parts), cost, ''.join(types), lengths, variant))
    out.sort(key=lambda q: q.cost)
    return out


def body_valid(course, g, pts: np.ndarray, pad: float) -> bool:
    """course.bodyClear(p, c, pad) for every pose (V4 check, vectorised)."""
    if not len(pts):
        return False
    return bool(course.body_clear_many(pts[:, 0], pts[:, 1], pts[:, 2], g, pad).all())


def plan(start, goal, course, g, pad: float = 0.003, step: float = 0.006) -> Tuple[Optional[RSPath], List[RSPath]]:
    """V4 CarbotRS.plan: every candidate checked (for the GUI audit), the shortest
    collision-free one wins. -> (winner or None, evaluated)."""
    allc = candidates(start, goal, g.r, step)
    if allc:
        P = np.vstack([q.points for q in allc])
        ok = course.body_clear_many(P[:, 0], P[:, 1], P[:, 2], g, pad)
        i = 0
        for q in allc:
            n = len(q.points)
            q.valid = bool(ok[i:i + n].all())
            q.reject = '' if q.valid else 'Footprint leaves drivable area'
            i += n
    winner = next((q for q in allc if q.valid), None)
    return winner, allc


def to_path4(points: np.ndarray) -> np.ndarray:
    """N x 5 -> N x 4 [x, y, a, dir] (the planning Path convention)."""
    return np.asarray(points, float)[:, :4].copy()
